crates/forge/src/block_number_map.rs

```diff
@@ -0,0 +1,118 @@
+        &mut self,
+        url: String,
+        hash: Felt252,
+    ) -> Option<&BlockNumber> {
+        self.url_and_hash_to_block_number.get(&(url, hash))
+    }
+
+    #[must_use]
+    pub fn get_url_to_latest_block_number(&self) -> &HashMap<String, BlockNumber> {
+        &self.url_to_latest_block_number
+    }
+}
+
+async fn get_latest_block_number(url: &Url) -> Result<BlockNumber> {
+    let client = JsonRpcClient::new(HttpTransport::new(url.clone()));
+
+    match Handle::current()
+        .spawn(async move { client.get_block_with_tx_hashes(BlockId::Tag(Latest)).await })
+        .await?
+    {
+        Ok(MaybePendingBlockWithTxHashes::Block(block)) => Ok(BlockNumber(block.block_number)),
+        _ => Err(anyhow!("Could not get the latest block number".to_string())),
+    }
+}
+
+async fn get_block_number_from_hash(url: &Url, block_hash: &Felt252) -> Result<BlockNumber> {
+    let client = JsonRpcClient::new(HttpTransport::new(url.clone()));
+
+    let hash = BlockId::Hash((*block_hash).clone().into_());
+    match Handle::current()
+        .spawn(async move { client.get_block_with_tx_hashes(hash).await })
+        .await?
+    {
+        Ok(MaybePendingBlockWithTxHashes::Block(block)) => Ok(BlockNumber(block.block_number)),
+        _ => Err(anyhow!(format!(
+            "Could not get the block number for block with hash 0x{}",
+            block_hash.to_str_radix(16)
+        ))),
+    }
+}
+
+pub async fn validated_fork_config_from_fork_params(
+    fork_params_string: &RawForkParams,
+    block_number_map: &mut BlockNumberMap,
+) -> Result<ValidatedForkConfig> {
+    let url_str = fork_params_string.url.clone();
+    let url = fork_params_string.url.parse()?;
+    let block_number = match fork_params_string.block_id_type.to_lowercase().as_str() {
+        "number" => BlockNumber(fork_params_string.block_id_value.parse()?),
+        "hash" => {
+            let block_hash =
+                Felt252::from(fork_params_string.block_id_value.parse::<BigInt>().unwrap());
+            if let Some(block_number) =
+                block_number_map.get_block_number_for_hash(url_str.clone(), block_hash.clone())
+            {
+                *block_number
+            } else {
+                let block_number = get_block_number_from_hash(&url, &block_hash).await?;
+                block_number_map.add_block_number_for_hash(url_str, block_hash, block_number);
+                block_number
+            }
+        }
+        "tag" => {
+            assert_eq!(fork_params_string.block_id_value, "Latest");
+            if let Some(block_number) = block_number_map.get_latest_block_number(&url_str) {
+                *block_number
+            } else {
+                let latest_block_number = get_latest_block_number(&url).await?;
+                block_number_map.add_latest_block_number(url_str, latest_block_number);
+                latest_block_number
+            }
+        }
+        _ => unreachable!(),
```

Comment:
Can we program this so it does not use `unreachable!` macro?

Reply:
I can throw a panic here or add an error message, but idk why it changes anything. Those are verified in scarb test collector or when parsing config from Scarb.toml